lightningd/invoice.c

```diff
@@ -213,13 +213,16 @@ static bool hook_gives_failcode(struct log *log,
 		fatal("Invalid invoice_payment_hook failure_code: %.*s",
 		      toks[0].end - toks[1].start, buffer);
 
-	/* UPDATE isn't valid for final nodes to return, and I think
-	 * we assert elsewhere that we don't do this! */
-	if (val & UPDATE)
-		fatal("Invalid invoice_payment_hook UPDATE failure_code: %.*s",
-		      toks[0].end - toks[1].start, buffer);
-	*failcode = val;
-	return true;
+	/* FIXME: Allow hook to return its own failmsg directly? */
+	if (val == WIRE_TEMPORARY_NODE_FAILURE)
+		return towire_temporary_node_failure(ctx);
+	if (val != WIRE_INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS)
+		log_broken(hin->key.channel->log,
+			   "invoice_payment hook returned failcode %u,"
+			   " changing to incorrect_or_unknown_payment_details",
+			   val);
+
+	return failmsg_incorrect_or_unknown(ctx, hin);
```

Comment:
These look reasonable and cover all the plugins I'm aware of :+1:

Seems a bit strange though to structure it like this `if (positive) {} if (negative) {} positive`